System Information and Drive Details into database for status and user edit

microtechno9000 opened the pull request:
# Description

Added some new features into ARM
- System Information is pulled from the database rather than pulling from the OS on each page load. Memory and temperature values are still pulled each load to keep current. Although now pulling from a class, intention being if a future update/PR wanted to graph temperature or memory over time the update should be smaller.
- System drives (CD/DVD/Bluray) are automatically scanned and added to the database and settings page. Scanning determines the drive type, i.e. CD, Bluray or DVD to the user.
- Drives report current job, with link to the job, the previous job and the current open state with an svg graphic.
- Users can open/close specific drives from the settings configuration page. Note: this does not tie into the current ARM UI abandon job tasking, but will not process the eject if a job is in progress.

Fixes # (issue title here)
- PR has been merged into the current v2_devel, with merge changes assessed and resolved
- Fixed ARM handling of database migrations. Appeared on each loading of the home/index page a lengthy check was automatically being done. During testing this caused at times 10's of arm.db files to be created. The fix in this PR notifies the user that the database needs updating, creates timestamped backup and migrates to the new version all from the ARM UI.
- System settings pages have been broken out into separate pages within the code for easy management. refer templates/settings folder with each of the tab html code blocks.

Known Issue
- When running ARM within a docker container or VM, the system information (CPU and Memory usage) has been noted to report 0% on some systems.

Comment
At present, a lot of functionality is captured within the db.models class structure, and as functions within the ui.utils. In developing this, I would have liked to take some of the functionality and move it into a class style structure to allow easier growth and management, but without rewriting a lot of functions within ARM I have tried to meet in the middle ground.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?
Tested with TV Series DVD, no issues in running through ARM. The changes updated the job id to the new system drive table.
On finish, loading the settings page validated the job had finished and moved the completed job to the previous field.
Tested multiple times with an old arm.db (alembic head f1054468c1c7), with issues resolved and caught.
Only one change has been made to the arm ripper functionality, to report the job id to the respective drive. Otherwise most changes are present within the arm ui.

- [ ] Ubuntu 18.04
- [x] Ubuntu 20.04
- [ ] Debian 10
- [ ] Debian 11
- [ ] Docker
- [ ] Other (Please state here)

# Checklist:

- [x] **I have submitted this PR against the `v2_devel` branch (Unless the main branch has a security issue)**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have tested that my fix is effective or that my feature works

microtechno9000 commented:
ready for review, fixed codesmells
the bug and other codesmells are existing code within ARM.

1337-server commented:
Can you fix this, so it passes the quality checks ?

microtechno9000 commented:
> Can you fix this, so it passes the quality checks ?

sure can, havent had much time lately to fix. will have a look

1337-server commented:
> > Can you fix this, so it passes the quality checks ?
> 
> sure can, havent had much time lately to fix. will have a look

Awesome, thanks!

1337-server commented:
You will also need to update the submodule for this, It's still pointing to the old one.

shitwolfymakes commented:
Bump. Happy to review and merge once everything is squared away with the quality checks. Don't forget to sync with the updates I merged a few minutes ago

microtechno9000 commented:
> Bump. Happy to review and merge once everything is squared away with the quality checks. Don't forget to sync with the updates I merged a few minutes ago

code smells and Quality Checks have been fixed after many a small edit! Finally worked out what the checks were doing and installed pycodestyle to run the check locally

shitwolfymakes commented:
Love this. I'll try to review later today.

shitwolfymakes commented:
@microtechno9000 looks like there's some issues with the quality of the checks. Once that's remediated request a review from me

shitwolfymakes commented:
> At present, a lot of functionality is captured within the db.models class structure, and as functions within the ui.utils. In developing this, I would have liked to take some of the functionality and move it into a class style structure to allow easier growth and management, but without rewriting a lot of functions within ARM I have tried to meet in the middle ground.

That will help in some cases but don't worry about that too much. Most of the functionality here that is a floating function wouldn't much help being made a member of a class. The fact that most of these are the equivalent to static functions in a compiled language is perfectly fine.

That being said, organization of said functions is still a bit messy. Slowly but surely we're chipping away at the cruft, and your contribution is an enormous help in that effort!

microtechno9000 commented:
> @microtechno9000 looks like there's some issues with the quality of the checks. Once that's remediated request a review from me

not sure which sorry, all I can see are the checks that fail the arm install but can't see the specific issue they are failing on.

microtechno9000 commented:
@shitwolfymakes removed the sqlalchemy.orm import, solved the missing QA check issue. I left that in, as I thought I needed it for the models, as I am using relationships. Appears to have been pulled in with the db class variable.
There are two checks still not passing, but not sure why

- [Test A.R.M UI (3.6.0)](https://github.com/automatic-ripping-machine/automatic-ripping-machine/actions/runs/3320578879/jobs/5487125713#logs) - failing on the install ARM dependencies, not sure what this check is doing, or what the error is reporting
- [Test A.R.M UI (3.9)](https://github.com/automatic-ripping-machine/automatic-ripping-machine/actions/runs/3320578879/jobs/5487125951#logs) - failing on run ARM UI, with the below error. I haven't made any change to the ui_settings migration '9cae4aa05dd7_create_settingsui_table', that was done in an earlier PR. I have added later migrations, using alembic which should link all the files to maintain a migrate-able database??
`sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: ui_settings`

Sorry about all the back and forth. This is my first foray into using github and first time using python.

shitwolfymakes commented:
No worries, the quality checks tend to be quite finicky. We'll get through this

shitwolfymakes commented:
Here is the error log:
```Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1808, in _execute_context
    self.dialect.do_execute(
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 732, in do_execute
    cursor.execute(statement, parameters)
sqlite3.OperationalError: no such table: ui_settings

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/runner/work/automatic-ripping-machine/automatic-ripping-machine/./arm/runui.py", line 10, in <module>
    import arm.ui.routes  # noqa E402
  File "/home/runner/work/automatic-ripping-machine/automatic-ripping-machine/arm/../arm/ui/routes.py", line 29, in <module>
    armui_cfg = models.UISettings.query.get(1)
  File "<string>", line 2, in get
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/util/deprecations.py", line 401, in warned
    return fn(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 943, in get
    return self._get_impl(ident, loading.load_on_pk_identity)
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/orm/query.py", line 947, in _get_impl
    return self.session._get_impl(
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 2892, in _get_impl
    return db_load_fn(
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/orm/loading.py", line 530, in load_on_pk_identity
    session.execute(
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 1692, in execute
    result = conn._execute_20(statement, params or {}, execution_options)
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1620, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 325, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1487, in _execute_clauseelement
    ret = self._execute_context(
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1851, in _execute_context
    self._handle_dbapi_exception(
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2032, in _handle_dbapi_exception
    util.raise_(
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1808, in _execute_context
    self.dialect.do_execute(
  File "/opt/hostedtoolcache/Python/3.10.8/x64/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 732, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: ui_settings
[SQL: SELECT ui_settings.id AS ui_settings_id, ui_settings.use_icons AS ui_settings_use_icons, ui_settings.save_remote_images AS ui_settings_save_remote_images, ui_settings.bootstrap_skin AS ui_settings_bootstrap_skin, ui_settings.language AS ui_settings_language, ui_settings.index_refresh AS ui_settings_index_refresh, ui_settings.database_limit AS ui_settings_database_limit 
FROM ui_settings 
WHERE ui_settings.id = ?]
[parameters: (1,)]
(Background on this error at: https://sqlalche.me/e/14/e3q8)
Error: Process completed with exit code 1.
```

microtechno9000 commented:
@shitwolfymakes I think I am finally done, with it all resolved :)
The check was throwing an exception because no arm.db file existed on initilisation, and there is a global UISetting declared in the routes.py. Might have broken that in changing some of the db setup checks elsewhere.
Should be good now, added in a check and then db creation if no file exists.
Could be neater, with some more global functions/checks to avoid code repetition, but as you have noted that is future hopefully non-breaking work.

shitwolfymakes commented:
Glad to hear! This is why the checks are key 👍

Will try to review this evening

shitwolfymakes commented:
In the meantime, if you're able to test the docket container that will build from these changes and post log files from successful runs, that'd be extremely helpful.

microtechno9000 commented:
> In the meantime, if you're able to test the docket container that will build from these changes and post log files from successful runs, that'd be extremely helpful.

havent created a container before, us there a script or guide?

shitwolfymakes commented:
Have a look at our [wiki](https://github.com/automatic-ripping-machine/automatic-ripping-machine/wiki/Building-ARM-docker-image-from-source) guide. This will be a great opportunity to find out if they're still up to date.

In the `git clone ...` command, you will need to replace the URL with that of your own repo

1337-server commented:
The code smell was a 'TODO' that was flagged, but this can be ignored.

shitwolfymakes commented:
We're almost ready to start final testing on this, just need #607 to get merged, then you'll be able to update to use the fixed dependencies

shitwolfymakes commented:
@microtechno9000 When fixing the dependencies merge conflicts:
1) in your copy of the repo, delete `rm -rf .git/modules/arm-dependencies/` (otherwise git wont be able to update the submodule)
2) choose "Accept Incoming" or "Accept Theirs". Your copies of `arm-dependencies/` and `.gitmodules` should be identical to ours

microtechno9000 commented:
> 1. rm -rf .git/modules/arm-dependencies/

Removed the directory and merged the changes from the v2_devel update accepting the changes into this repo

shitwolfymakes commented:
Awesome, now it's time to test the docker container. Go ahead and follow the instruction on [this](https://github.com/automatic-ripping-machine/automatic-ripping-machine/wiki/Building-ARM-docker-image-from-source) wiki page, and attach a log file from a successfully completed rip

If the documentation is out of date, let us know and we'll work on fixing it.

microtechno9000 commented:
> Awesome, now it's time to test the docker container. Go ahead and follow the instruction on [this](https://github.com/automatic-ripping-machine/automatic-ripping-machine/wiki/Building-ARM-docker-image-from-source) wiki page, and attach a log file from a successfully completed rip
> 
> If the documentation is out of date, let us know and we'll work on fixing it.

@shitwolfymakes I am out for the next few weeks, so won't be able to test the docker container and provide any feedback, earliest would be early December.

shitwolfymakes commented:
Duly noted, I'm also testing it myself.

 I am getting error after starting it without a db (I deleted the old one). It creates one but to throws an exception `Error: 'coretemp'`.

Additionally, this may have created a folder in `/home/arm` named `0777``

microtechno9000 commented:
oh, I added a bunch of functionality to allow for migrating old databases to a new ARM version, but didnt touch the old code much for creating a new DB. If you have the error log send it through and I will see if its a quick fix.

microtechno9000 commented:
@shitwolfymakes found the bug, routes.py was not calling the right function on start. Calling the right one now and will then create the database
Also found a bug with the coretemp, when the sensor is not supported it bugs out, added a len = check for the returned dict

shitwolfymakes commented:
Fix allow me to access the UI, and UI appears to work with some issues. All issues listed here are observed in my testing of the docker container, with the following conditions:
- Running with and without the `--cpuset-cpus` flag
- No database file before container launch
- No previous rips or logs

1. Under `Arm Settings > General Info > Disk Drives`, there are no drives listed
2. The System Info widget does not populate with any data other than the defaults
3. The System Memory widget does not display the total amount of system memory
4. Container successfully initiates ripping when a disk is inserted, but endlessly loops.
    1. Once the job starts looping, the job cannot be successfully abandoned. The syslog looks like this:
```
asdf
```

Attached is a copy of the logfile.

microtechno9000 commented:
I haven't tested within docker, haven't set that up before. I will have to park this PR for a few weeks to investigate and fix.

microtechno9000 commented:
> Fix allow me to access the UI, and UI appears to work with some issues. All issues listed here are observed in my testing of the docker container, with the following conditions:
> 
>     * Running with and without the `--cpuset-cpus` flag
> 
>     * No database file before container launch
> 
>     * No previous rips or logs
> 
> 
>     1. Under `Arm Settings > General Info > Disk Drives`, there are no drives listed
> 
>     2. The System Info widget does not populate with any data other than the defaults
> 
>     3. The System Memory widget does not display the total amount of system memory
> 
>     4. Container successfully initiates ripping when a disk is inserted, but endlessly loops.
>        
>        1. Once the job starts looping, the job cannot be successfully abandoned, the only fix is killing the container, removing the dvd, and deleting the database file. The log looks like this:
> 
> 
> ```
> Nov  3 00:40:53 626177bf4a77 ARM: Entering docker wrapper
> Nov  3 00:40:53 626177bf4a77 ARM: [ARM] Starting ARM for DVD on sr0
> [2022-11-03 00:40:55,731] DEBUG ARM: json_api.process_logfile active
> [2022-11-03 00:40:55,731] DEBUG ARM: json_api.process_logfile using handbrake
> [2022-11-03 00:40:55,736] DEBUG ARM: json_api.process_handbrake_logfile Cant find index
> [2022-11-03 00:40:55,744] DEBUG ARM: json_api.get_x_jobs couldn't get config
> [2022-11-03 00:40:55,744] DEBUG ARM: json_api.get_x_jobs jobs  - we have 1 jobs
> [2022-11-03 00:40:57,733] DEBUG ARM: json_api.process_logfile active
> [2022-11-03 00:40:57,733] DEBUG ARM: json_api.process_logfile using handbrake
> [2022-11-03 00:40:57,737] DEBUG ARM: json_api.process_handbrake_logfile Cant find index
> [2022-11-03 00:40:57,739] DEBUG ARM: json_api.get_x_jobs couldn't get config
> [2022-11-03 00:40:57,739] DEBUG ARM: json_api.get_x_jobs jobs  - we have 1 jobs
> ```
> 
> Attached is a copy of the rip job logfile. [POINT_BREAK.log](https://github.com/automatic-ripping-machine/automatic-ripping-machine/files/9924995/POINT_BREAK.log)

@shitwolfymakes I have updated the PR with a fix for some of the issues, although having issues with docker.
The below are now fixed, the database was not initialing correctly. fixed that so the system settings and other widgets are displaying data in docker and my local instance correctly
1. Under Arm Settings > General Info > Disk Drives, there are no drives listed
2. The System Info widget does not populate with any data other than the defaults
3. The System Memory widget does not display the total amount of system memory

Had issues getting docker to work, the [wiki instructions ](https://github.com/automatic-ripping-machine/automatic-ripping-machine/wiki/Building-ARM-docker-image-from-source)are missing a line in the run config, need to add in the automatic-ripping-machine at the end for the docker image name.

```
...
--name "automatic-ripping-machine" \
   automatic-ripping-machine
```

4. Not sure how to resolve this one. I can't get the logs from the docker container to output to the folder I want.
```
   -v "/home/arm/docker/logs:/home/arm/logs" \
   -v "/home/arm/docker/config:/etc/arm/config" \
```

Tried modifying the armui.service file, but that does not create any logs within the /home/arm/docker/logs folder.
Permissions are working, as the config directory has the arm config files.
Looks like you have it working on your machine? What config have you got set. I haven't created containers before, so might be missing something simple.

microtechno9000 commented:
@shitwolfymakes on issue 4, it appears there is an issue with HandbrakeCLI. Doing some testing I am unable to resolve it, and the issue occurs with both os_vals and v2_devel. I have have gone through 'handbrake.py' but can't see where the issue resides, not sure if it is something that this PR has introduced or something in one of the recent PR's, would need @1337-server to answer that, I am not across this code enough to understand what it is doing.
I haven't run this in docker yet, as the below didn't work on my server, so didn't see any  point testing in docker.

os_vals branch (INDEPENDENCE_DAY_166952349977.log)
```
[11-27-2022 15:02:45] INFO ARM: arm_ripper.start_transcode ************* Starting Transcode With HandBrake *************
[11-27-2022 15:02:45] DEBUG ARM: arm_ripper.start_transcode handbrake_all: /dev/sr0, /home/arm/media/transcode/unidentified/INDEPENDENCE_DAY_166952349977, /home/arm/logs/INDEPENDENCE_DAY_166952349977.log
[11-27-2022 15:02:45] INFO ARM: utils.sleep_check_process Transcode limit is disabled
[11-27-2022 15:02:45] INFO ARM: handbrake.handbrake_all Starting BluRay/DVD transcoding - All titles
[11-27-2022 15:02:45] INFO ARM: handbrake.get_track_info Using HandBrake to get information on all the tracks on the disc.  This will take a few minutes...
[11-27-2022 15:02:45] ERROR ARM: main.<module> A fatal error has occurred and ARM is exiting.  See traceback below for details.
[11-27-2022 15:02:45] DEBUG ARM: utils.database_adder Trying to add Notifications
[11-27-2022 15:02:45] DEBUG ARM: utils.database_adder successfully written Notifications to the database
[11-27-2022 15:02:45] DEBUG ARM: models.eject Unmounted disc /dev/sr0
[11-27-2022 15:02:49] DEBUG ARM: models.eject Failed to eject /dev/sr0
```

v2_devel (INDEPENDENCE_DAY_166952509809.log)

```
[11-27-2022 15:29:25] INFO ARM: arm_ripper.start_transcode ************* Starting Transcode With HandBrake *************
[11-27-2022 15:29:25] INFO ARM: utils.sleep_check_process Transcode limit is disabled
[11-27-2022 15:29:25] INFO ARM: handbrake.handbrake_all Starting BluRay/DVD transcoding - All titles
[11-27-2022 15:29:25] INFO ARM: handbrake.get_track_info Using HandBrake to get information on all the tracks on the disc.  This will take a few minutes...
[11-27-2022 15:29:25] DEBUG ARM: handbrake.get_track_info Sending command: HandBrakeCLI -i /dev/sr0 -t 0 --scan
[11-27-2022 15:29:45] DEBUG ARM: utils.put_track Track #01 Length: 8836 fps: 25.000 aspect: 1.78 Mainfeature: False Source: HandBrake
[11-27-2022 15:29:45] DEBUG ARM: utils.database_adder Trying to add Track
[11-27-2022 15:29:45] DEBUG ARM: utils.database_adder successfully written Track to the database
[11-27-2022 15:29:45] DEBUG ARM: utils.put_track Track #02 Length: 8332 fps: 25.000 aspect: 1.78 Mainfeature: True Source: HandBrake
[11-27-2022 15:29:45] DEBUG ARM: utils.database_adder Trying to add Track
[11-27-2022 15:29:45] DEBUG ARM: utils.database_adder successfully written Track to the database
[11-27-2022 15:29:45] DEBUG ARM: utils.put_track Track #04 Length:   44 fps: 25.000 aspect: 1.78 Mainfeature: False Source: HandBrake
[11-27-2022 15:29:45] DEBUG ARM: utils.database_adder Trying to add Track
[11-27-2022 15:29:45] DEBUG ARM: utils.database_adder successfully written Track to the database
[11-27-2022 15:29:45] DEBUG ARM: handbrake.handbrake_all Total number of tracks is None
[11-27-2022 15:29:45] INFO ARM: handbrake.handbrake_all Processing track #1 of None. Length is 8836 seconds.
[11-27-2022 15:29:45] ERROR ARM: main.<module> A fatal error has occurred and ARM is exiting.  See traceback below for details.
[11-27-2022 15:29:45] DEBUG ARM: utils.database_adder Trying to add Notifications
[11-27-2022 15:29:45] DEBUG ARM: utils.database_adder successfully written Notifications to the database
[11-27-2022 15:29:45] DEBUG ARM: models.eject Unmounted disc /dev/sr0
[11-27-2022 15:29:50] DEBUG ARM: models.eject Failed to eject /dev/sr0
```

shitwolfymakes commented:
So on your version, handbrake shits the bed as soon as it tries to access the disk. Have you tried cleaning the disk and trying to re-run?

Also, in your devtools scripts, there are new packages required. Please freeze the version numbers for all new packages in your requirements.txt, and document them in a changelog section of your PR description. Additionally, please provide a README in that folder, showcasing how to use them. Thanks!

shitwolfymakes commented:
Also, I cannot login to arm in the docker container or the desktop installation, even after deleting the database. Log below:

shitwolfymakes commented:
> Also, I cannot login to arm in the docker container or the desktop installation, even after deleting the database. Log below:
> 
> ```
> [2022-11-30 21:10:09,487] INFO ARM: runui.<module> Starting ARMUI on interface address - 172.17.0.3:8080
> [2022-11-30 21:10:47,137] DEBUG ARM: routes.login No admin found
> [2022-11-30 21:11:27,732] DEBUG ARM: routes.login No admin found
> [2022-11-30 21:11:27,734] DEBUG ARM: routes.login user= None
> [2022-11-30 21:11:27,734] DEBUG ARM: routes.handle_exception Error: 'NoneType' object has no attribute 'password'
> ```

Ignore this. Looks like my dev environment got hilariously borked at some point. No issues with logins after restoring

shitwolfymakes commented:
Database version check occurs every time `/index.html` loads. This is unnecessary, and should run each time on startup

1337-server commented:
> Database version check occurs every time `/index.html` loads. This is unnecessary, and should run each time on startup instead

This was already in ARM and not part of the PR


---


> @shitwolfymakes on issue 4, it appears there is an issue with HandbrakeCLI. Doing some testing I am unable to resolve it, and the issue occurs with both os_vals and v2_devel. I have have gone through 'handbrake.py' but can't see where the issue resides, not sure if it is something that this PR has introduced or something in one of the recent PR's, would need @1337-server to answer that, I am not across this code enough to understand what it is doing. I haven't run this in docker yet, as the below didn't work on my server, so didn't see any point testing in docker.

> ```
> [11-27-2022 15:29:25] INFO ARM: arm_ripper.start_transcode ************* Starting Transcode With HandBrake *************
> [11-27-2022 15:29:25] INFO ARM: utils.sleep_check_process Transcode limit is disabled
> [11-27-2022 15:29:25] INFO ARM: handbrake.handbrake_all Starting BluRay/DVD transcoding - All titles
> [11-27-2022 15:29:25] INFO ARM: handbrake.get_track_info Using HandBrake to get information on all the tracks on the disc.  This will take a few minutes...
> [11-27-2022 15:29:25] DEBUG ARM: handbrake.get_track_info Sending command: HandBrakeCLI -i /dev/sr0 -t 0 --scan
> [11-27-2022 15:29:45] DEBUG ARM: utils.put_track Track #01 Length: 8836 fps: 25.000 aspect: 1.78 Mainfeature: False Source: HandBrake
> [11-27-2022 15:29:45] DEBUG ARM: utils.database_adder Trying to add Track
> [11-27-2022 15:29:45] DEBUG ARM: utils.database_adder successfully written Track to the database
> [11-27-2022 15:29:45] DEBUG ARM: utils.put_track Track #02 Length: 8332 fps: 25.000 aspect: 1.78 Mainfeature: True Source: HandBrake
> [11-27-2022 15:29:45] DEBUG ARM: utils.database_adder Trying to add Track
> [11-27-2022 15:29:45] DEBUG ARM: utils.database_adder successfully written Track to the database
> [11-27-2022 15:29:45] DEBUG ARM: utils.put_track Track #04 Length:   44 fps: 25.000 aspect: 1.78 Mainfeature: False Source: HandBrake
> [11-27-2022 15:29:45] DEBUG ARM: utils.database_adder Trying to add Track
> [11-27-2022 15:29:45] DEBUG ARM: utils.database_adder successfully written Track to the database
> [11-27-2022 15:29:45] DEBUG ARM: handbrake.handbrake_all Total number of tracks is None
> [11-27-2022 15:29:45] INFO ARM: handbrake.handbrake_all Processing track #1 of None. Length is 8836 seconds.
> [11-27-2022 15:29:45] ERROR ARM: main.<module> A fatal error has occurred and ARM is exiting.  See traceback below for details.
> [11-27-2022 15:29:45] DEBUG ARM: utils.database_adder Trying to add Notifications
> [11-27-2022 15:29:45] DEBUG ARM: utils.database_adder successfully written Notifications to the database
> [11-27-2022 15:29:45] DEBUG ARM: models.eject Unmounted disc /dev/sr0
> [11-27-2022 15:29:50] DEBUG ARM: models.eject Failed to eject /dev/sr0
> ```

This happens when HandBrake finds the info for the disc, but returns it to handbrake.handbrake_all malformed or errors out when returning the data. In this case, it seems to be returning nothing, even though there has been tracks found.

shitwolfymakes commented:
> > Database version check occurs every time `/index.html` loads. This is unnecessary, and should run each time on startup instead
> 
> This was already in ARM and not part of the PR

Interesting. I don't remember seeing it in the syslog. I've also noticed that the `feed_json` logging that used to run every 10 milliseconds is no longer occurring

shitwolfymakes commented:
Hmm. When I refresh the home page during a job, it is stuck loading. I assume that has something to do `json_api.process_logfile` running almost 100 times a second?

For context, I hit refresh and it still isn't done refreshing almost 2 minutes later

1337-server commented:
> > > Database version check occurs every time `/index.html` loads. This is unnecessary, and should run each time on startup instead
> > 
> > 
> > This was already in ARM and not part of the PR
> 
> Interesting. I don't remember seeing it in the syslog. I've also noticed that the `feed_json` logging that used to run every 10 milliseconds is no longer occurring

I could be misremembering, but I'm sure it should be pinging every few seconds, not ms. so if your database/settings is set to '10' this is way too low and will cause crashes/hangs.
I usually set it to ping every 2000

shitwolfymakes commented:
> I could be misremembering, but I'm sure it should be pinging every few seconds, not ms. so if your database/settings is set to '10' this is way too low and will cause crashes/hangs. I usually set it to ping every 2000

Interesting. The default setting is 10, which is absolutely insane. It should be updated to 2000 by default with a note that the lower it is, the higher the chance of hangs and instability.
- I just tried updating the setting from the UI. After saving changes, refreshing or navigating to a different page results in the following error: ``

shitwolfymakes commented:
Okay. Upping the delay appears to not only significantly speed up the UI's responsiveness, but it also fixed the database connection error I kept gett

@microtechno9000 Would you kindly update that the default index refresh interval to 2000 as part of this PR?

microtechno9000 commented:
Resolved the issue, the alembic database file was loading in default values for the ARMUI settings which had 10ms, I was then initialising the db with a second line at 2000ms, but the ARMUI only grabs the first line.
Resolved by fixing the alembic file to 2000ms and removing the initilisation code I had when there is no DB.
to check you will need to remove the current database, you can now do this with the below. I added that in and tested it when fixing the DB issue

`python3 devtools/armdevtools.py -db_rem`

shitwolfymakes commented:
I just now noticed this, but you seem to be adding a new `.github/workflows/docker-publish.yml` file. This is unnecessary, as no `docker` branch exists, and the file duplicates the functionality of our `.github/workflows/publish-image.yml`. 

On another note, I figured out why that workflow is failing. You need to create a DockerHub account, and add the following as Actions secrets to your repository:
`DOCKERHUB_USERNAME`: Your dockerhub username
`DOCKERHUB_TOKEN`: A personal access token docker requires you to generate to use for CLI connections (Store this in a password manager or other secure storage!)

Sorry there's been a lack of documentation on this, we've made a lot of radical changes since the last time the wiki was thoroughly updated. I'm going to rewrite large swathes of it once 2.7 is released.

microtechno9000 commented:
@shitwolfymakes  the docker file was pulled in from an early merge that I did from a @1337-server fork. When I started the change, I forked a 1337 branch, then re-based it to arm v2_devel, so that might be why. Not sure what that file does, so you can remove if required.

on the requirements.txt file, I can't update it, tried adding the below for the devtools, but get an error when trying to add the file change to git.

```
git add arm-dependencies/requirements.txt 
fatal: Pathspec 'arm-dependencies/requirements.txt' is in submodule 'arm-dependencies'
```
rows to add to requirements.txt

```
argparse==1.4.0
..
colorama=0.4.4
```

microtechno9000 commented:
@shitwolfymakes I am not sure how to setup and modify the docker-publish.yml file, if you have creds and an API key, are you able to add them into this branch/PR?

shitwolfymakes commented:
> @shitwolfymakes I am not sure how to setup and modify the docker-publish.yml file, if you have creds and an API key, are you able to add them into this branch/PR?

That's the beauty of it. I wrote the workflow to be fork-agnostic, so all you have to do is set up those secrets in your fork and it'll run!

shitwolfymakes commented:
Docker container run was successful with a Blu-ray. See attached log
[MOULIN_ROUGE_167134539271.log](https://github.com/automatic-ripping-machine/automatic-ripping-machine/files/10254513/MOULIN_ROUGE_167134539271.log)

shitwolfymakes commented:
Docker container run was successful with a DVD. See attached log
[POINT_BREAK.log](https://github.com/automatic-ripping-machine/automatic-ripping-machine/files/10258318/POINT_BREAK.log)

Will review the code tomorrow. The code for CPU temp and memory utilization still doesn't seem to work properly, but that can be dealt with after the heavy lifting of merging this is handled.

microtechno9000 commented:
> Ubuntu 20.04 thick-client run was successful with a DVD. See attached log [POINT_BREAK.log](https://github.com/automatic-ripping-machine/automatic-ripping-machine/files/10258318/POINT_BREAK.log)
> 
> Will review the code tomorrow. The code for CPU temp and memory utilization still doesn't seem to work properly, but that can be dealt with after the heavy lifting of merging this is handled.

that's weird, I tested with docker and the system values came through.  when you get a moment could you send me a copy of the logs and system info, will confirm with my server

shitwolfymakes commented:
The cpu usage is either 0 or 100% whenever I refresh the page, and the memory usage is totally static. Granted, that could be a result of me running the base OS in a VM instead of bare metal. Additionally,y refresh interval is 2000ms

shitwolfymakes commented on .github/workflows/docker-publish.yml:
```diff
@@ -0,0 +1,37 @@
+name: publish-docker-image
```
Remove this file please and thanks

microtechno9000 replied:
done

shitwolfymakes commented:
@microtechno9000 please update to the latest commit of `arm-dependencies`

shitwolfymakes commented on arm/migrations/versions/9cae4aa05dd7_create_settingsui_table.py:
```diff
@@ -36,7 +36,7 @@ def upgrade():
                        'bootstrap_skin': "spacelab",
                        'language': "en",
-                       'index_refresh': 10,
+                       'index_refresh': 2000,
```
This is gonna save a lot of frustration for users, thanks for tracking it down! Please make sure the associated tooltip info is updated to reflect the new value

microtechno9000 commented:
> @microtechno9000 please update to the latest commit of `arm-dependencies`

Think I managed to get it updated as a sub-module. If it is still out of date, how do I go about updating it?

shitwolfymakes commented:
> > @microtechno9000 please update to the latest commit of `arm-dependencies`
> 
> 
> 
> Think I managed to get it updated as a sub-module. If it is still out of date, how do I go about updating it?

Should just be `git submodule update --remote`

You can confirm it's up-to-date by checking the `@ hash against the latest commit to the `arm-dependencies repo`

shitwolfymakes commented on arm/models/models.py:
```diff
@@ -491,3 +493,138 @@ def get_d(self):
         return return_dict
+
+
+class SystemInfo(db.Model):
```
The rule should be one class per file, but we can push that refactor to another time

microtechno9000 replied:
agree, the file at the moment is rather large and cumbersome 
if that is the general rule I can reactor with any other changes I make against v2_devel

shitwolfymakes replied:
It's not formalized yet, so you're good here. A refactor that big needs to be it's own PR

shitwolfymakes commented on arm/ui/routes.py:
```diff
@@ -670,51 +704,30 @@ def home():
+    global page_support_databaseupdate
 
+    # Force a db update
+    # Microtechno9000 comment:
```
Remove these, they'll still be viewable in the git history

shitwolfymakes commented on arm/ui/utils.py:
```diff
@@ -79,42 +83,176 @@ def check_db_version(install_path, db_file):
+                                 f"Head is {head_revision} and database is {db_version}.  Exiting arm.")
+
+
+def arm_alembic_get():
```
In the future, it might be a good idea to factor these functions out into a `db_utils` file

microtechno9000 replied:
I am keen to move the database handler to a class and call it from the ui and ripper to avoid duplication of code. However, in developing this change was not sure on how the project was setup and the ripper code looked to hard to update without breaking everything.
How do we keep track of this? I can't add a task to the 2.7 project area

shitwolfymakes commented:
Blu-Ray rip using the docker container ran successfully. Log attached
[MOULIN_ROUGE.log](https://github.com/automatic-ripping-machine/automatic-ripping-machine/files/10333996/MOULIN_ROUGE.log)

shitwolfymakes commented:
DVD rip using the thick client install ran successfully. Log attached
[POINT_BREAK.log](https://github.com/automatic-ripping-machine/automatic-ripping-machine/files/10334191/POINT_BREAK.log)

shitwolfymakes commented:
@microtechno9000 Please update the first post to include a section titled "Known Issues". List that the CPU and Memory widgets don't always properly populate values.

Once that's done, I'll complete the review and merge!

shitwolfymakes reviewed (approved):
I can't begin to describe what an amazing job you've done with this. Great fucking job!
